pkg/scheduler/scheduler_test.go

```diff
@@ -792,6 +796,420 @@ func TestPatchScheduleResultForResourceBinding(t *testing.T) {
+							{Name: "cluster1", Replicas: 1},
+						},
+					}, nil
+				},
+			}
+
+			eventBroadcaster := record.NewBroadcaster()
+			eventRecorder := eventBroadcaster.NewRecorder(testScheme, corev1.EventSource{Component: "test-scheduler"})
+
+			s := &Scheduler{
+				KarmadaClient:        fakeClient,
+				queue:                queue,
+				bindingLister:        bindingLister,
+				clusterBindingLister: clusterBindingLister,
+				Algorithm:            mockAlgo,
+				eventRecorder:        eventRecorder,
+			}
+
+			s.queue.Add(tc.key)
+
+			if tc.shutdown {
+				s.queue.ShutDown()
+			}
+
+			result := s.scheduleNext()
+
+			assert.Equal(t, tc.expectResult, result, "scheduleNext return value mismatch")
+
+			if !tc.shutdown {
+				assert.Equal(t, 0, s.queue.Len(), "Queue should be empty after processing")
+			}
+		})
+	}
+}
+
+func TestPlacementChanged(t *testing.T) {
+	tests := []struct {
+		name                 string
+		placement            *policyv1alpha1.Placement
+		appliedPlacementStr  string
+		observedAffinityName string
+		want                 bool
+	}{
+		{
+			name: "placement changed",
+			placement: &policyv1alpha1.Placement{
+				ClusterAffinity: &policyv1alpha1.ClusterAffinity{
+					ClusterNames: []string{"cluster1", "cluster2"},
+				},
+			},
+			appliedPlacementStr:  `{"clusterAffinity":{"clusterNames":["cluster1"]}}`,
+			observedAffinityName: "",
+			want:                 true,
+		},
+		{
+			name: "placement not changed",
+			placement: &policyv1alpha1.Placement{
+				ClusterAffinity: &policyv1alpha1.ClusterAffinity{
+					ClusterNames: []string{"cluster1", "cluster2"},
+				},
+			},
+			appliedPlacementStr:  `{"clusterAffinity":{"clusterNames":["cluster1","cluster2"]}}`,
+			observedAffinityName: "",
+			want:                 false,
+		},
+		{
+			name: "invalid applied placement string",
+			placement: &policyv1alpha1.Placement{
+				ClusterAffinity: &policyv1alpha1.ClusterAffinity{
+					ClusterNames: []string{"cluster1", "cluster2"},
+				},
+			},
+			appliedPlacementStr:  `invalid json`,
+			observedAffinityName: "",
+			want:                 false,
+		},
+		{
+			name:                 "empty placement",
+			placement:            &policyv1alpha1.Placement{},
+			appliedPlacementStr:  `{}`,
+			observedAffinityName: "",
+			want:                 false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// Test for ClusterResourceBinding
+			crb := &workv1alpha2.ClusterResourceBinding{
+				Spec: workv1alpha2.ResourceBindingSpec{
+					Placement: tt.placement,
+				},
+				Status: workv1alpha2.ResourceBindingStatus{
+					SchedulerObservedAffinityName: tt.observedAffinityName,
+				},
+			}
+			got := placementChanged(*crb.Spec.Placement, tt.appliedPlacementStr, crb.Status.SchedulerObservedAffinityName)
+			assert.Equal(t, tt.want, got, "placementChanged() for ClusterResourceBinding result mismatch")
+
+			// Test for ResourceBinding
+			rb := &workv1alpha2.ResourceBinding{
```

Comment:
rb and crb should be no different here, is one enough to test?

Reply:
Yes, they have identical behavior.